tokio-executor/src/thread_pool/thread_pool.rs

```diff
@@ -0,0 +1,97 @@
+use crate::thread_pool::{shutdown, Builder, JoinHandle, Spawner};
+use crate::Executor;
+
+use std::fmt;
+use std::future::Future;
+
+/// Work-stealing based thread pool for executing futures.
+pub struct ThreadPool {
+    spawner: Spawner,
+
+    /// Shutdown waiter
+    shutdown_rx: shutdown::Receiver,
+}
+
+impl ThreadPool {
+    /// Create a new ThreadPool with default configuration
+    pub fn new() -> ThreadPool {
+        Builder::new().build()
+    }
+
+    pub(super) fn from_parts(spawner: Spawner, shutdown_rx: shutdown::Receiver) -> ThreadPool {
+        ThreadPool {
+            spawner,
+            shutdown_rx,
+        }
+    }
+
+    /// Returns reference to `Spawner`.
+    ///
+    /// The `Spawner` handle can be cloned and enables spawning tasks from other
+    /// threads.
+    pub fn spawner(&self) -> &Spawner {
+        &self.spawner
+    }
+
+    /// Spawn a task
+    pub fn spawn<F>(&self, future: F) -> JoinHandle<F::Output>
+    where
+        F: Future + Send + 'static,
+        F::Output: Send + 'static,
+    {
+        self.spawner.spawn(future)
+    }
+
+    /// Spawn a task in the background
+    pub(crate) fn spawn_background<F>(&self, future: F)
+    where
+        F: Future<Output = ()> + Send + 'static,
+    {
+        self.spawner.spawn_background(future);
+    }
+
+    /// Block the current thread waiting for the future to complete.
+    ///
+    /// The future will execute on the current thread, but all spawned tasks
+    /// will be executed on the thread pool.
+    pub fn block_on<F>(&self, future: F) -> F::Output
+    where
+        F: Future,
+    {
+        crate::global::with_threadpool(self, || {
+            let mut enter = crate::enter()
+                .ok()
+                .expect("attempting to block while on a Tokio executor");
+            enter.block_on(future)
+        })
+    }
+
+    /// Shutdown the thread pool.
+    pub fn shutdown_now(&mut self) {
+        if self.spawner.workers().close() {
+            self.shutdown_rx.wait();
+        }
+    }
+}
+
+impl Executor for &'_ ThreadPool {
```

Comment:
```suggestion
impl Executor for &ThreadPool {
```
nit: lifetime can be elided